### PR TITLE
Base: make CFLock_t a pthread mutex on Linux, FreeBSD

### PR DESCRIPTION
**Pull request by compnerd**
Unify the behaviour across macOS and Linux, FreeBSD to use
pthread_mutex_t when CFLock is used.  This will allow us to remove
references to pthread_mutex_{lock,unlock} which makes CoreFoundation
easier to port to Windows.

### CODE REVIEW

**Comment by compnerd** [comment]
CC: @mikeash @millenomi

**Comment by compnerd** [comment]
@swift-ci please test

File: CoreFoundation/Base.subproj/CFInternal.h [hunk 1]
```diff
@@ -540,29 +540,24 @@ CF_INLINE Boolean __CFLockTry(volatile CFLock_t *lock) {
...
-#define OSSpinLockUnlock(lock) __CFUnlock(lock)
+#define __CFLock(LP) ({ (void)pthread_mutex_lock(LP); })
+#define __CFUnlock(LP) ({ (void)pthread_mutex_unlock(LP); })
+#define __CFLockTry(LP) ({ (void)pthread_mutex_trylock(LP) == 0; })
+
+// NOTE: we cannot use pthread_spinlock_t here as there is no static initializer
+// defined for them in the POSIX specification.
+typedef int32_t OSSpinLock;
+#define OS_SPINLOCK_INIT 0
+#define OSSpinLockLock(lock) ({ while (__sync_val_compare_and_swap(lock, 0, -0)) sleep(0); })
```
Comment (spevans) [inline CoreFoundation/Base.subproj/CFInternal.h]: Is `-0` correct? Should it be `~0` as above?

Comment (compnerd) [reply CoreFoundation/Base.subproj/CFInternal.h]: Bah!  Reading is hard.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by spevans** [comment]
@swift-ci test

**Comment by compnerd** [comment]
I think that I am going to just fold this into the larger PR for the windows port.

**Pull request closed without merging**